logicalplan/distribute.go

```diff
@@ -99,7 +126,37 @@ func (m DistributedExecutionOptimizer) Optimize(plan parser.Expr) parser.Expr {
 	return plan
 }
 
-func (m DistributedExecutionOptimizer) makeSubQueries(current *parser.Expr, engines []api.RemoteEngine) Coalesce {
+func getRemoteAggregation(aggr *parser.AggregateExpr, engines []api.RemoteEngine) parser.Expr {
+	groupingSet := make(map[string]struct{})
+	for _, lbl := range aggr.Grouping {
+		groupingSet[lbl] = struct{}{}
+	}
+
+	for _, engine := range engines {
+		for _, lbls := range engine.LabelSets() {
+			for _, lbl := range lbls {
+				if aggr.Without {
+					delete(groupingSet, lbl.Name)
+				} else {
+					groupingSet[lbl.Name] = struct{}{}
+				}
+			}
+		}
+	}
+
+	groupingLabels := make([]string, 0, len(groupingSet))
+	for lbl := range groupingSet {
+		groupingLabels = append(groupingLabels, lbl)
+	}
+
+	sort.Strings(groupingLabels)
+
+	remoteAggregation := *aggr
```

Comment:
Nit: maybe this can be simplified to:
```go
	aggr.Grouping = groupingLabels
	return aggr
```

I think the result should be the same.

Reply:
That would modify the input parameter which is used for the root aggregation. To avoid that, `remoteAggregation` should be a copy instead.